### PR TITLE
Add poppler/20.09.0 recipe

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **poppler/20.09.0**

- [x] Requires #2798
- [x] Requires #3299
---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

Fixes https://github.com/conan-io/wishlist/issues/194

### CODE REVIEW

File: recipes/poppler/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        "enable_libcurl": [True, False],
+        "enable_zlib": [True, False],
+        "use_float": [True, False],
+
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+        "fontconfiguration": "generic",
+        "enable_cairo": False,
```
Comment (uilianries) [inline recipes/poppler/all/conanfile.py:43]: Interesting, Does Poppler have an internal Cairo implementation?

File: recipes/poppler/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        self._cmake = CMake(self)
+
+        self._cmake.definitions["ENABLE_UNSTABLE_API_ABI_HEADERS"] = False
+        self._cmake.definitions["BUILD_GTK_TESTS"] = False
+        self._cmake.definitions["BUILD_QT5_TESTS"] = False
+        self._cmake.definitions["BUILD_QT6_TESTS"] = False
+        self._cmake.definitions["BUILD_CPP_TESTS"] = False
+
+        self._cmake.definitions["ENABLE_UTILS"] = False
+        self._cmake.definitions["ENABLE_CPP"] = True
```
Comment (uilianries) [inline recipes/poppler/all/conanfile.py:142]: Is there any reason for not keep cpp as optional?

Comment (madebr) [reply recipes/poppler/all/conanfile.py:142]: Not that I think of.
Making it optional, and on by default would be ok too.
I just didn't bother here.

File: recipes/poppler/all/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/poppler/all/conanfile.py:43]: No, this should require cairo. This is a bug.

**Comment by Croydon** [comment]
Please add `Fixes https://github.com/conan-io/wishlist/issues/194`

File: recipes/poppler/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+            return "none"
+        else:
+            return str(self.options.with_libjpeg)
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+
+        self._cmake.definitions["ENABLE_UNSTABLE_API_ABI_HEADERS"] = False
```
Comment (madebr) [inline recipes/poppler/all/conanfile.py:147]: FIXME:
on MSVC, these options are not recognized:
```
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    WITH_FONTCONFIGURATION
    WITH_GLIB
    WITH_GObjectIntrospection
    WITH_GTK
```

Comment (madebr) [reply recipes/poppler/all/conanfile.py:147]: Fixed.

`WITH_GTK` is still emitted because it is only used when the glib option is enabled.

**Comment by Vectrom** [comment]
Have you ever tried to build it with qt (e.g. from bincrafters)?

I am trying it, but on MSVC on Windows I receives error:
```poppler/20.09.0@vectrom/stable: ERROR: Package '1fe33c4fc444ec8ba6393d5f8e753bd36bc3a9d0' build failed
poppler/20.09.0@vectrom/stable: WARN: Build folder C:\Users\runneradmin\.conan\data\poppler\20.09.0\vectrom\stable\build\1fe33c4fc444ec8ba6393d5f8e753bd36bc3a9d0
ERROR: poppler/20.09.0@vectrom/stable: Error in build() method, line 209
	cmake.build()
	ConanException: Error 1 while executing cmake --build C:\Users\runneradmin\.conan\data\poppler\20.09.0\vectrom\stable\build\1fe33c4fc444ec8ba6393d5f8e753bd36bc3a9d0\build_subfolder --config Debug -- /m:2 /verbosity:minimal
```

MacOS and Ubuntu passed successfully (without -with_qt option also Windows passed successfully).

**Comment by madebr** [comment]
> Have you ever tried to build it with qt (e.g. from bincrafters)?

No I haven't.
The error message you posted is quite short.
It should print more warnings/errors.

**Comment by Vectrom** [comment]
[1_Win (1).txt](https://github.com/conan-io/conan-center-index/files/5472250/1_Win.1.txt)

This is a full log (from my test repository with this recipe and Github Actions, but locally effect is the same). I don't see other errors.

**Comment by madebr** [comment]
Thanks for the complete log.
The only error I could find is:
```
2020-11-01T23:10:49.7766355Z C:\Users\runneradmin\.conan\data\poppler\20.09.0\vectrom\stable\build\70059836dd0e2cd21d744e9eaa4fb611f2235aa5\build_subfolder\source_subfolder\qt5\src\poppler-qt5_autogen\include_Release\EWIEGA46WW/moc_poppler-optcontent.cpp(66,83): error C2491: 'Poppler::OptContentModel::staticMetaObject': definition of dllimport static data member not allowed (compiling source file C:\Users\runneradmin\.conan\data\poppler\20.09.0\vectrom\stable\build\70059836dd0e2cd21d744e9eaa4fb611f2235aa5\build_subfolder\source_subfolder\qt5\src\poppler-qt5_autogen\mocs_compilation.cpp) [C:\Users\runneradmin\.conan\data\poppler\20.09.0\vectrom\stable\build\70059836dd0e2cd21d744e9eaa4fb611f2235aa5\build_subfolder\source_subfolder\qt5\src\poppler-qt5.vcxproj]
```
I cannot deduce from the log what options you used, but I suppose you used the default `poppler:shared=False`?
(I recommend to set [`log.print_run_commands = True`](https://docs.conan.io/en/latest/reference/conanfile/other.html#running-commands) in your `conan.conf` for logging what is executed through `self.run`)

Can you try building with `-o poppler:shared=True`?
Looking at `poppler-qt5`'s [`poppler-export.h`](https://github.com/freedesktop/poppler/blob/master/qt5/src/poppler-export.h) file, it looks like it does not support static builds.
Some patching is required.

**Comment by Vectrom** [comment]
> I cannot deduce from the log what options you used, but I suppose you used the default `-o poppler:shared=False` with `-o poppler:with_qt=True`?

Yes.

Logs for options `-o poppler:shared=True -o poppler:with_qt=True' below:
[1_Win (1).txt](https://github.com/conan-io/conan-center-index/files/5472487/1_Win.1.txt)

I see problems with linking now (unresolved symbols).

**Comment by madebr** [comment]
@Vectrom 
The log actually contained only one (link) error:
```
2020-11-02T02:28:32.9176646Z poppler-annotation.obj : error LNK2001: unresolved external symbol "public: static double const AnnotFreeText::undefinedFontPtSize" (?undefinedFontPtSize@AnnotFreeText@@2NB) [C:\Users\runneradmin\.conan\data\poppler\20.09.0\vectrom\stable\build\c49bc8cdb3e7fe244e222c52cf20ceaf06708543\build_subfolder\source_subfolder\qt5\src\poppler-qt5.vcxproj]
```
But my `CMAKE_WIN32_EXPORT_ALL_SYMBOLS` hack cannot fix this issue.
I've opened https://gitlab.freedesktop.org/poppler/poppler/-/merge_requests/672 to fix it thorougly.

I've marked my merge request as a draft because, to get the tests working, I had to also export `GooString` in [`goo/GooString.h`](https://gitlab.freedesktop.org/poppler/poppler/-/blob/master/goo/GooString.h).
This makes it fail for for MSVC because `GooString` extends `std::string` which causes the linker to complain about duplicate `std::string` symbols.

**Comment by Vectrom** [comment]
I'll checked version with changes from PR. Package compiles correctly for static and with only one error on Windows for shared (mentioned earlier problem with GooString). 
Thanks!

File: recipes/poppler/all/conandata.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  "20.09.0":
+    url: "https://poppler.freedesktop.org/poppler-20.09.0.tar.xz"
+    sha256: "4ed6eb5ddc4c37f2435c9d78ff9c7c4036455aea3507d1ce8400070aab745363"
+patches:
+  "20.09.0":
+    - patch_file: "patches/0001-cmake-use-PROJECT_SOURCE_DIR-fix-CMAKE_MODULE_PATH.patch"
+      base_path: "source_subfolder"
+    - patch_file: "patches/0002-cmake-use_CONAN_PKG-targets.cmake"
+      base_path: "source_subfolder"
+    - patch_file: "patches/0003-cmake-disable-tests.cmake"
```
Comment (prince-chrismc) [inline recipes/poppler/all/conandata.yml:11]: please rename to `.patch`

**Review by prince-chrismc — CHANGES_REQUESTED** [review]
~typo~

File: recipes/poppler/all/conandata.yml [hunk 5]
```diff
@@ -6,7 +6,7 @@ patches:
   "20.09.0":
     - patch_file: "patches/0001-cmake-use-PROJECT_SOURCE_DIR-fix-CMAKE_MODULE_PATH.patch"
       base_path: "source_subfolder"
-    - patch_file: "patches/0002-cmake-use_CONAN_PKG-targets.cmake"
+    - patch_file: "patches/0002-cmake-use_CONAN_PKG-targets.patch"
       base_path: "source_subfolder"
     - patch_file: "patches/0003-cmake-disable-tests.cmake"
```
Comment (prince-chrismc) [inline recipes/poppler/all/conandata.yml:11]: +1

Comment (madebr) [reply recipes/poppler/all/conandata.yml:11]: oh didn't see that one

Comment (prince-chrismc) [reply recipes/poppler/all/conandata.yml:11]: you did not rename the file(s) either

Comment (madebr) [reply recipes/poppler/all/conandata.yml:11]: In my defence, I did `0002-cmake-use_CONAN_PKG-targets.patch` :smile:

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Review by jgsogo — APPROVED** [review]

**Pull request merged**